src/reconcile/rec_col.c

```diff
@@ -1653,12 +1675,40 @@ __wt_rec_col_var(
     }
 
     /* If we were tracking a record, write it. */
-    if (rle != 0)
+    if (rle != 0) {
+        if (!last.deleted)
+            wrote_real_values = true;
         WT_ERR(
           __rec_col_var_helper(session, r, salvage, last.value, &last.tw, rle, last.deleted, NULL));
+    }
+
+    /*
+     * If we have not generated any real values but only deleted-value cells, bail out and call the
+     * page empty instead. (Note that this should always be exactly one deleted-value cell, because
+     * of the RLE handling, so we can't have split.) This will create a namespace gap like those
+     * produced by truncate.
+     *
+     * Skip this step if:
+     *     - We're in salvage, to avoid any odd interactions with the way salvage reconstructs the
+     *       namespace.
+     *     - There were invisible updates, because then the page isn't really empty. Also, at least
+     *       for now if we try to restore updates to an empty page col_modify will trip on its
+     *       shoelaces.
+     */
+    if (!wrote_real_values && salvage == NULL && r->leave_dirty == false) {
+        WT_ASSERT(session, r->entries == 1);
+        r->entries = 0;
+        /* Don't bother adjusting r->space_avail or r->first_free. */
+    }
 
     /* Write the remnant page. */
-    ret = __wt_rec_split_finish(session, r);
+    WT_ERR(__wt_rec_split_finish(session, r));
+
+    /*
+     * Clear the split recno so that if we reuse the append list again later after a truncate, the
+     * old bound on it doesn't come back to life and cause confusion and consternation.
+     */
+    page->modify->mod_col_split_recno = WT_RECNO_OOB;
```

Comment:
Should this happen regardless of an error? i.e: would this move down into the error handling logic?

Reply:
I evidently thought not at the time. Hmm... ah, see line 1546 (and 876). It's possible that the VLCS copy of that logic is no longer needed because we're tolerating gaps now. That's the only thing it's actually used for (besides asserting about it in col_modify) ... all the same, removing it makes me nervous and I'd want to run a bunch more stress tests. Let me chew on this a bit too. Among other things it would eliminate the problem with slow reconciliation and gaps that I've been worrying about.

Reply:
OK, thanks. I'm inclined not to open another can of worms. This was mostly a curiosity question.. I'm happy with the status quo if you are.

Reply:
I think it is the right thing to do (that is, don't use mod_col_split_recno for VLCS) -- the problem I was worried about is that if page A splits into page A and page B, and then page B gets deleted and discarded before A is reconciled (this is ~impossible to set up on purpose AFAICT, but I know no reason to suppose it's _impossible_) the next insert into B's namespace will come back to A and trigger a false positive in the mod_col_split_recno check in col_modify.

Disabling the hook entirely loses the crosscheck there, but also avoids that concern, and leaving a gap at the end of A no longer matters.

It works so far, and there's no reason it should create a problem. It's just that everything related to splits and races in splits is nervous-making, so I want to run more stress tests before merging it.

I will push the change in a bit.

Reply:
Pushed in [4df09f3](https://github.com/wiredtiger/wiredtiger/pull/8129/commits/4df09f33d27ec72dab2e99f55ced1a727b18ba48).